components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java

```diff
@@ -77,10 +77,12 @@ public Origin getOrigin() {
             Connection connection = dequeue();
             if (connection != null) {
                 borrowedCount.incrementAndGet();
+                sink.onCancel(() -> availableConnections.add(connection));
                 sink.success(connection);
             } else {
                 if (waitingSubscribers.size() < poolSettings.maxPendingConnectionsPerHost()) {
-                    this.waitingSubscribers.add(sink.onDispose(() -> waitingSubscribers.remove(sink)));
+                    this.waitingSubscribers.add(sink);
+                    sink.onDispose(() -> waitingSubscribers.remove(sink));
```

Comment:
I think this change can be reverted. However, I thought we had an issue with the waitingSubscribers list and we haven't changed anything in that code path... Don't we need to change anything here?